### DISCUSSION TITLE
Support Mermaid in MDN Markdown

### DISCUSSION THREAD

**Discussion by Josh-Cena** [mdn/mdn-community#358] [Platform]
This is discussed with @schalkneethling on Slack.

[Mermaid](https://mermaid.js.org/intro/) is a tool for generating SVG diagrams from a DSL. GitHub has added support for it for just over a year: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

Why do we want to have it built-in? Because this generates the SVG markup within the HTML, instead of the traditional way of adding an `<img>` element that links to the SVG. The benefit is that the SVG text is selectable, and any `<a>` links within the SVG would be clickable. There doesn't seem to be a way to do this in MDN yet, without cluttering the MD source with a chunk of copy/pasted SVG.

Here's an SVG I'm developing:

![image](https://user-images.githubusercontent.com/55398995/230406580-bb99d0ab-098b-4522-a0e6-5d7fe1d00a06.png)

Where every box would be a clickable link to its class page.

**Comment by teoli2003** [comment]
I'm not against it: I dream of having a way to describe diagrams inside a document.

Nevertheless, it would be interesting to know where we could use it on MDN. Adding it to create one UML diagram doesn't seem like a good trade-off for the added complexity. I can think of a few other places this could be handy (like for network diagrams, handy in fetch, XHR, HTTP, …)

I have a few questions:
- How does this work wrt to accessibility? I guess we still need to create an alt text. How do we do this? Is the code generated accessible?
- Theme: does mermaid generate diagrams suited for different color schemes? Your example is on a bright white and is painful if I look at it in a dark room in dark mode.
- Are there some templates? We sure want to have every UML diagrams look the same (triangle sizes, relative dimensions…) or all network diagrams be similar? How is this done?

**Reply by Josh-Cena** [reply]
> where we could use it on MDN

We have a lot of diagrams on MDN, and yes, [network diagrams](https://mermaid.js.org/syntax/examples.html#sequencediagram-loops-alt-and-opt), [requirement diagrams](https://mermaid.js.org/syntax/requirementDiagram.html), [mindmaps](https://mermaid.js.org/syntax/mindmap.html) are all doable in Mermaid. Simply going for feature parity with GitHub is desirable as well, since writers will be familiar with it.

> wrt to accessibility

Diagrams are information-rich, so alt texts won't be able to describe them perfectly anyway. You can surely have aria titles, etc.; the Mermaid team is also working on it. See https://mermaid.js.org/config/accessibility.html

> Theme

This is, again, another big point of why we should have Mermaid built-in. By shipping more client code, we can dynamically toggle the theme based on color mode; otherwise, we can only ship one static embedded image. The above example looks like that exactly because I can only cater to one color mode with an `<img>` tag.

Here's one example from Docusaurus's integration: https://docusaurus.io/docs/markdown-features/diagrams

> Are there some templates?

Mermaid is really, really declarative and we can surely have one set of CSS files to unify everything. Here's the code for the diagram above:

~~~md
```mermaid
%%{
  init: {
    'theme': 'base',
    'themeVariables': {
      'primaryColor': 'white',
      'primaryTextColor': 'black',
      'primaryBorderColor': 'black',
      'lineColor': 'black'
    }
  }
}%%

classDiagram

`Object.prototype` <|-- `Iterator.prototype`
`Iterator.prototype` <|-- `GeneratorFunction.prototype.prototype`
`GeneratorFunction.prototype.prototype` <|-- `genFunc.prototype`
`genFunc.prototype` <|-- `genFunc()`

`Object.prototype` <|-- `Function.prototype`
`Function.prototype` <|-- `GeneratorFunction.prototype`
`GeneratorFunction.prototype` <|-- genFunc

`Function.prototype` <|-- Function
Function <|-- GeneratorFunction

genFunc <.. `genFunc()` : instanceof
`genFunc.prototype` "prototype" <-- genFunc
`GeneratorFunction.prototype.prototype` "prototype" <-- `GeneratorFunction.prototype`
`GeneratorFunction.prototype.prototype` --> "constructor" `GeneratorFunction.prototype`
`GeneratorFunction.prototype` "prototype" <-- GeneratorFunction
`GeneratorFunction.prototype` --> "constructor" GeneratorFunction
`Function.prototype` "prototype" <-- Function
`Function.prototype` --> "constructor" Function

class `Object.prototype` {
  __proto__

  hasOwnProperty()
  isPrototypeOf()
  propertyIsEnumerable()
  toLocaleString()
  toString()
  valueOf()
}
class `Iterator.prototype` {
  \[@@iterator]()
}
class `GeneratorFunction.prototype.prototype` {
  <<Generator.prototype>>
  [@@toStringTag] = "Generator"
  next()
  return()
  throw()
}

class genFunc["function* genFunc() {}"] {
  length = 0
  name = "genFunc"
}
class `Function.prototype` {
  \[@@hasInstance]()
  apply()
  bind()
  call()
}
class `GeneratorFunction.prototype` {
  [@@toStringTag] = "GeneratorFunction"
}
class GeneratorFunction {
  length = 1
  name = "GeneratorFunction"
}
class Function {
  length = 1
  name = "Function"
}

link `Object.prototype` "/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object"
link `Iterator.prototype` "/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator"
link `GeneratorFunction.prototype.prototype` "/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator"
link `Function.prototype` "/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function"
link `GeneratorFunction.prototype` "/en-US/docs/Web/JavaScript/Reference/Global_Objects/GeneratorFunction"
link GeneratorFunction "/en-US/docs/Web/JavaScript/Reference/Global_Objects/GeneratorFunction"
link Function "/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function"
```
~~~

**Comment by OnkarRuikar** [comment]
It looks like a nice to have feature. Mermaid has support for [font awesome style custom icons](https://mermaid.js.org/syntax/mindmap.html#icons) so many existing png, gpg files can be converted to the marmaid code. This will reduce repo code size a lot.

## Implementation approaches

We have three options to implement this in yari.
a. Github makes user agents render the diagrams(mermaid code to svg). They just load empty iframe with mermaid script after it's loaded the web page sends mermaid code to iframe. So the iframe renders it on its own.\
If javascript is disabled it won't work, so we can't do this.

b. Before doing normal builds in yari, pass the content md code through mermaid-cli.\
Which will convert mermaid code blocs to .svg files near the .md file and auto insert `![image](image.svg)` tags in markdown file.

c. Use .mmd(mermaid diagram) files like we use png, gif, svg etc. Convert them to svg during build.

Let me know if I've missed anything?

***

## Counter argument

Don't hate me for this.

Regarding strategy a. and b. above, current prod build is ~1hr long and dev build is ~1hr30min long. Processing mermaid code blocs will add to the build time a lot. When one day we'll have thousands of mermaid blocks all over the repos then this will become a real issue. We can ignore lengthy prod builds but longer dev build will really make working on yari a nightmare. We don't have true dev mode in yari ATM. :rage: 
Also, it's not just prod code we'll have to implement it in dev code(preview mode) as well. Which will be a significant effort depending on the strategy we choose.

Regarding adding complexity concern raised by @teoli2003.\
In order to handle linkable SVGs we can simply make yari render SVGs inline instead of `<img>` tags.\
Almost all of us are going to use [Mermaid live](https://mermaid.live/) to create/tinker diagrams first then copy the mmd code to .md file. So why not copy the generated svg instead?

I mean to keep the platform simple, we can always use Mermaid offline to generate those beautiful illustrations and use the generated SVGs in content.\
Apart from above discussion, is use of Mermaid just for ease of use or linkable/clickable svgs? what are other advantages?

**Comment by Josh-Cena** [comment]
Going back to this, @OnkarRuikar.

> If javascript is disabled it won't work, so we can't do this.

It's not a big deal—browser compatibility table doesn't work either, neither do all the interactive/live examples. Experience is degraded but no critical information is lost anyway. If you still don't like it let's look at the next option.

I personally don't have an opinion between 2 and 3, although I believe 2 works better than 3, because 3 is referencing a file that doesn't exist until build-time (I assume you still write `![](...svg)`, unless you invent some other reference syntax like `![](...mmd)` which I personally don't like), which means other Markdown renderers that potentially understand Mermaid won't be able to understand.

Regarding build time. Yari really needs an incremental build mode; I don't know how it does this, but doesn't Yari do partial builds in `yarn start` and PR preview? Also we can have persistent caching, etc., not a big problem either.

Really, I don't care if we support Mermaid _compilation_ in Yari, but I need to support Mermaid _rendering_, even if that means I need to compile Mermaid to SVG by hand. What I _don't_ want is having to copy the SVG source into Markdown just to make it selectable/clickable. If we can make `![](...svg)` automatically inlined as an SVG element, I would probably be happy for the time being.

> is use of Mermaid just for ease of use or linkable/clickable svgs? what are other advantages?

That's the major point yes. Apart from that, looking at the many existing custom diagrams created using various softwares:

- Uniform style.
- Light/dark mode support.
- Ease of maintenance without relying on a particular external software.
- Textual representation, less storage space required for the source (although the built size is the same).
- SVG is always superior to PNG/JPEG, which seems like most diagrams use.

**Reply by OnkarRuikar** [reply]
> It's not a big deal—browser compatibility table doesn't work either, neither do all the interactive/live examples. Experience is degraded but no critical information is lost anyway. If you still don't like it let's look at the next option.

[Pure CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/margin#examples) and HTML `{{EmbedLiveSample}}`s do work when JavaScript is disabled.\
Yes, no critical info is lost when BCD and example outputs are not shown. But imagine reader's frustration when critical diagram is missing from the discussion/documentation that refers it.

If top brass is willing to let go of "no JavaScript" support then we certainly can have Mermaids the way GitHub has.

***

> Regarding build time. Yari really needs an incremental build mode; I don't know how it does this, but doesn't Yari do partial builds in `yarn start` and PR preview? Also we can have persistent caching, etc., not a big problem either.

Refer the current [dev](https://github.com/mdn/yari/actions/workflows/dev-build.yml) and [prod](https://github.com/mdn/yari/actions/workflows/prod-build.yml) build times. I think prebuilding is done to get better response times and less load on server. As the built doc cache will reset when new content gets deployed everyday, we won't get the full benefit. Also, we can get away with low spec servers using prebuilt content.\
PR companion builds only modified files. And `yarn start` doesn't build anything deployable: browser keeps polling backend and backend sends content in JSON when contributor modifies the document.

**Reply by Josh-Cena** [reply]
> built doc cache will reset when new content gets deployed everyday, we won't get the full benefit

That was what my "persistent caching" is referring to. Even with serverless deployments I believe there is persistent caching options that we may want to look into.

**Reply by OnkarRuikar** [reply]
> That was what my "persistent caching" is referring to. Even with serverless deployments I believe there is persistent caching options that we may want to look into.

:thinking: One more solution could be to maintain a `payload` repo of built documents and simply checkout the repo on prod server after installing yari. And at the end of a day build and commit only the modified .md files in the `payload` repo and make prod server do `git clone` or `git pull`. With this daily deployments will happen in seconds.

In case of major changes in yari itself, we trigger our current full build and commit all the files to `payload` repo.

**Comment by bsmth** [comment]
See also https://github.com/mdn/shared-assets/issues/48 where Claas has suggested we build SVG on merges to `main` in that repo.
